### PR TITLE
DietPi-Software | Prometheus Node Exporter

### PR DESCRIPTION
**Pull request by alexiri**
This commit adds Prometheus Node Exporter, scratching my own itch as described in #5549. This is my first contact with this code base, so please let me know if I messed something up.

Note I've only been able to test this on a Raspberry Pi, it's the only model of SBC I have.

### CODE REVIEW

File: dietpi/dietpi-software [hunk 1]
```diff
@@ -3435,6 +3441,87 @@ _EOF_
...
+				'armv6l') arch='armv6';;
+				'armv7l') arch='armv7';;
+				'aarch64') arch='arm64';;
+				'x86_64') arch='amd64';;
+			esac
+
+			# Reinstall: Remove previous instances
+			[[ -e '/usr/local/bin/node_exporter' ]] && G_EXEC rm -R /usr/local/bin/node_exporter
+			[[ -e '/usr/local/bin/raspberrypi_exporter' ]] && G_EXEC rm -R /usr/local/bin/raspberrypi_exporter
+
```
Comment (MichaIng) [inline dietpi/dietpi-software]: The ne files overwrite the old ones, isn't it? So this explicit removal shouldn't be required, and skipping it better covers the case where the download fails, to not leave users without any executable:
```suggestion
```

File: dietpi/dietpi-software [hunk 2]
```diff
@@ -3435,6 +3441,87 @@ _EOF_
...
+			# Use the default port
+			local port="9100"
+
+			local arch=''
+			case $G_HW_ARCH_NAME in
+				'armv6l') arch='armv6';;
+				'armv7l') arch='armv7';;
+				'aarch64') arch='arm64';;
+				'x86_64') arch='amd64';;
+			esac
```
Comment (MichaIng) [inline dietpi/dietpi-software]: Satisfy shellcheck:
```suggestion
			case $G_HW_ARCH_NAME in
				'armv6l') local arch='armv6';;
				'armv7l') local arch='armv7';;
				'aarch64') local arch='arm64';;
				*) local arch='amd64';;
			esac
```

File: dietpi/dietpi-software [hunk 3]
```diff
@@ -3435,6 +3441,87 @@ _EOF_
...
 		fi
 
+		software_id=99 # Prometheus Node Exporter
+		if (( ${aSOFTWARE_INSTALL_STATE[$software_id]} == 1 )); then
+
+			Banner_Installing
+
+			# Use the default port
+			local port="9100"
+
```
Comment (MichaIng) [inline dietpi/dietpi-software]: Let's skip the variable as long as it's used in a single place only.
```suggestion
```

File: dietpi/dietpi-software [hunk 4]
```diff
@@ -3435,6 +3441,87 @@ _EOF_
...
+[Unit]
+Description=Node Exporter
+
+[Service]
+User=node_exporter
+ExecStart=/usr/local/bin/node_exporter --web.listen-address=":$port" --collector.textfile.directory /var/lib/node_exporter/textfile_collector
+
+[Install]
+WantedBy=multi-user.target
+_EOF_
```
Comment (MichaIng) [inline dietpi/dietpi-software]:
```suggestion
			cat << '_EOF_' > /etc/systemd/system/node_exporter.service
[Unit]
Description=Prometheus Node Exporter
Documentation=https://prometheus.io/docs/guides/node-exporter/

[Service]
User=node_exporter
ExecStart=/usr/local/bin/node_exporter --web.listen-address=9100 --collector.textfile.directory /var/lib/node_exporter/textfile_collector

[Install]
WantedBy=multi-user.target
_EOF_
```

File: dietpi/dietpi-software [hunk 5]
```diff
@@ -3435,6 +3441,87 @@ _EOF_
...
+			# Download binaries
+			Download_Install $(curl -sSfL 'https://api.github.com/repos/prometheus/node_exporter/releases/latest' | mawk -F\" "/\"browser_download_url\": \".*node_exporter-.*.linux-$arch.tar.gz\"/{print \$4}")
+			G_EXEC mv node_exporter*/node_exporter /usr/local/bin/
+			G_EXEC chmod +x /usr/local/bin/node_exporter
+
+			# User
+			Create_User node_exporter
+
+			# Working directory
+			G_EXEC mkdir -p /var/lib/node_exporter/textfile_collector
```
Comment (MichaIng) [inline dietpi/dietpi-software]: Shall we use `/mnt/dietpi_userdata/node_exporter` as data directory, to align with other software options and use external drive in many cases?

We could also use `/opt/node_exporter` for the binaries then, in case it is not usually intended to be called from console by times.

File: dietpi/dietpi-software [hunk 6]
```diff
@@ -3435,6 +3441,87 @@ _EOF_
...
+[Service]
+User=node_exporter
+ExecStart=/usr/local/bin/node_exporter --web.listen-address=":$port" --collector.textfile.directory /var/lib/node_exporter/textfile_collector
+
+[Install]
+WantedBy=multi-user.target
+_EOF_
+
+			aSTART_SERVICES+=('node_exporter')
+
```
Comment (MichaIng) [inline dietpi/dietpi-software]: I guess this shall be started by `dietpi-services`? In this case, simply add it to the list there, which makes `aSTART_SERVICES` obsolete here. Same for `raspberrypi_exporter`.
```suggestion
```

File: dietpi/dietpi-software [hunk 7]
```diff
@@ -3435,6 +3441,87 @@ _EOF_
...
+			# If we're on a platform with vcgencmd, we can use raspberrypi_exporter as well
+			if [[ $(command -v vcgencmd) != "" ]]; then
```
Comment (MichaIng) [inline dietpi/dietpi-software]: Use exit code instead of checking the output.
```suggestion
			if command -v vcgencmd > /dev/null; then
```
We could also use `(( $G_HW_MODEL < 10 ))` to check whether it's an RPi, in which case `vcgencmd` should always be present?

File: dietpi/dietpi-software [hunk 8]
```diff
@@ -3435,6 +3441,87 @@ _EOF_
...
+[Install]
+WantedBy=timers.target
+_EOF_
+
+				aSTART_SERVICES+=('raspberrypi_exporter')
+
+			fi
+
+			G_DIETPI-NOTIFY 2 "Prometheus Node Exporter serving metrics at: http://$HOSTNAME:$port/metrics"
+		fi
```
Comment (MichaIng) [inline dietpi/dietpi-software]: Ah, there `port` was used a second time. Well, let's leave this info for the docs, at this point in the middle of an install output most users wouldn't see it anyway 😉.
```suggestion
		fi
```
I thought about showing such info accumulated as very last `dietpi-software` output, for all installed software titles, but that would be a dedicated PR.

File: dietpi/dietpi-software [hunk 9]
```diff
@@ -13537,6 +13624,31 @@ If no WireGuard (auto)start is included, but you require it, please do the follo
...
 
+		software_id=99 # Prometheus Node Exporter
+		if (( ${aSOFTWARE_INSTALL_STATE[$software_id]} == -1 )); then
+
+			Banner_Uninstalling
+			if [[ -f '/etc/systemd/system/raspberrypi_exporter.service' ]]
+			then
+				G_EXEC systemctl disable --now raspberrypi_exporter.service
+				G_EXEC rm /etc/systemd/system/raspberrypi_exporter.service
+			fi
```
Comment (MichaIng) [inline dietpi/dietpi-software]: Also remove possible service file edits:
```suggestion
			fi
			[[ -d '/etc/systemd/system/raspberrypi_exporter.service.d' ]] && G_EXEC rm -R /etc/systemd/system/raspberrypi_exporter.service.d
```
Same for `node_exporter.service`.

File: dietpi/dietpi-software [hunk 10]
```diff
@@ -3435,6 +3441,87 @@ _EOF_
...
+				'aarch64') arch='arm64';;
+				'x86_64') arch='amd64';;
+			esac
+
+			# Reinstall: Remove previous instances
+			[[ -e '/usr/local/bin/node_exporter' ]] && G_EXEC rm -R /usr/local/bin/node_exporter
+			[[ -e '/usr/local/bin/raspberrypi_exporter' ]] && G_EXEC rm -R /usr/local/bin/raspberrypi_exporter
+
+			# Download binaries
+			Download_Install $(curl -sSfL 'https://api.github.com/repos/prometheus/node_exporter/releases/latest' | mawk -F\" "/\"browser_download_url\": \".*node_exporter-.*.linux-$arch.tar.gz\"/{print \$4}")
```
Comment (MichaIng) [inline dietpi/dietpi-software]: Satisfy shellcheck:
```suggestion
			Download_Install "$(curl -sSfL 'https://api.github.com/repos/prometheus/node_exporter/releases/latest' | mawk -F\" "/\"browser_download_url\": \".*node_exporter-.*.linux-$arch.tar.gz\"/{print \$4}")"
```

File: dietpi/dietpi-software (hunk 7 above)
Comment (alexiri) [reply dietpi/dietpi-software]: I wasn't sure if only RPis have `vcgencmd`, so I figured it was best to check for the binary itself, which is what I really care about.

File: dietpi/dietpi-software (hunk 5 above)
Comment (alexiri) [reply dietpi/dietpi-software]: This is a binary, so I wouldn't really call it userdata. It won't be called from the console ever, this is just a daemon serving metrics over http. I'm happy to move it to `/opt/node_exporter` if you prefer.

File: dietpi/dietpi-software (hunk 6 above)
Comment (alexiri) [reply dietpi/dietpi-software]: After removing this, the services aren't started automatically after install. What's the right way of doing this without `aSTART_SERVICES`?

File: dietpi/dietpi-software (hunk 5 above)
Comment (MichaIng) [reply dietpi/dietpi-software]: Still suggesting to also move `/var/lib/node_exporter/textfile_collector` to `/mnt/dietpi_userdata/node_exporter`, so we are consistent with other software options 🙂.

File: dietpi/dietpi-software [hunk 11]
```diff
@@ -3435,6 +3441,79 @@ _EOF_
...
+User=node_exporter
+ExecStart=/opt/node_exporter/node_exporter --web.listen-address=":9100" --collector.textfile.directory /var/lib/node_exporter/textfile_collector
+
+[Install]
+WantedBy=multi-user.target
+_EOF_
+
+			# If we're on a platform with vcgencmd, we can use raspberrypi_exporter as well
+			if command -v vcgencmd > /dev/null; then
+				Download_Install "https://github.com/fahlke/raspberrypi_exporter/archive/master.zip"
```
Comment (MichaIng) [inline dietpi/dietpi-software]: Let's save a little download bandwidth:
```suggestion
				Download_Install 'https://github.com/fahlke/raspberrypi_exporter/archive/master.tar.gz'
```

File: dietpi/dietpi-software (hunk 5 above)
Comment (alexiri) [reply dietpi/dietpi-software]: That folder isn't really userdata, though, it's more like a temporary file. A single file is created in that directory by `raspberrypi_exporter` and will be overwritten every 15 minutes. The contents include cpu frequency stats, temperatures, voltages, etc. That file will be served by `node_exporter` along with the other metrics it gathers.

I can move it to userdata if you really want, but I don't think it really belongs there, the file is useless outside of the particular point in time it's generated.

Comment (MichaIng) [reply dietpi/dietpi-software]: Would it then make sense to store it in `/run/node_exporter` instead and add `RuntimeDirectory=node_exporter` to `node_exporter.service`? So we don't do disk I/O for a temporary file.

The question is only how to configure `raspberrypi_exporter` to write to this directory instead, and also we need to document it, if `/var/lib/node_exporter` is the default and ... okay no chance: https://github.com/fahlke/raspberrypi_exporter#prerequisites
> node_exporter `--collector.textfile.directory` must be set to `/var/lib/node_exporter/textfile_collector`

Only thing we could do is using a symlink.

Comment (MichaIng) [reply dietpi/dietpi-software]: Made a related feature request: https://github.com/fahlke/raspberrypi_exporter/issues/4

Comment (MichaIng) [reply dietpi/dietpi-software]: Ah, it's a bash script with `METRICS_FILE` which can be easily adjusted.

**Review by MichaIng — DISMISSED** [review]

**Comment by MichaIng** [comment]
The timer btw somehow runs only every minute at `:34` here, while according to the `OnCalendar` value it should run every 15 seconds 🤔. However the latter sounds overkill to me anyway. Shall we adjust it to explicitly run every minute?

Otherwise everything works pretty well here, tested on VMs and RPi.

**Comment by MichaIng** [comment]
@alexiri 
Do you agree with the change? Little downside is that if `node_exporter` is stopped while `raspberrypi_exporter.timer` is still running, it throws errors since `/run/node_exporter` doesn't exist anymore. I wonder whether we should add some condition in this case to skip `raspberrypi_exporter` in this case.

**Comment by MichaIng** [comment]
Added a condition now to skip the RPi exporter gracefully if the directory does not exist. I'll merge now, we can fine tune when required.

**Review by MichaIng — APPROVED** [review]

**Comment by alexiri** [comment]
Looks great, thanks @MichaIng!
